src/main/scala/strawman/collection/mutable/ArrayDeque.scala

```diff
@@ -165,14 +165,12 @@ class ArrayDeque[A] private(var array: Array[AnyRef], var start: Int, var end: I
 
   override def reverse = {
     val r = new ArrayDeque[A](initialSize = size)
-    var i = length
-    while(0 < i) {
-      i -= 1
-      r.appendAssumingCapacity(this(i))
-    }
+    this.foreach(r.prependAssumingCapacity)
```

Comment:
In addition to @Ichoran 's comment, having your own foreach loop provides the compiler with an additional place to profile, and as the JIT does use-site profiling and loop profiling on a case by case basis, a dedicated one often is faster than a re-used/shared one.  Whether that is true depends on the order of inlining versus profiling.    The closure, with Scala's 2.12+ encoding has a better chance at being just as fast, as the JVM gets better at optimizing for lambdas.   For now, the safe thing is your own loop to avoid the lambda, but in the long run it might not matter.